### ISSUE TITLE
Autostart at Logon only opens a note

### ISSUE THREAD

**Issue by aguador** [tomboy-notes/tomboy-ng#111]
Running 0.24 on MX Linux (XFCE 4.12.3), 64-bit fully up to date, Spanish language interface.

To reproduce:

1. Open tomboy-ng Settings
2. In first tab check "Autostart at Logon"
3. Close session / reboot
4. Sign in

Expected behaviour: tomboy-ng starts with splash screen and/or in system tray

Actual behaviour: A new note is opened (as if it were in single note mode) and tomboy-ng icon does not appear in the system tray. Closing the note quits tomboy-ng.

ps aux | grep tomboy shows: "tomboy-ng %f" running, as opposed to "tomboy-ng" on Mageia 8

(tomboy-ng.desktop DOES have: EXEC=tomboy-ng %f -- on both machines)

I would also note that the dark theme is not properly picked up in the new note opened, as in #110.

**Comment by aguador** [comment]
Additional information: on closing or elimination the new (blank) note, the splash appears momentarily as tomboy-ng quits.

**Comment by davidbannon** [comment]
Messy !  The "%f" in the desktop file is there to allow things like your file manager to open a note when its double clicked (assuming you have associated tomboy-ng with that sort of file). Clearly thats not useful in a tomboy-ng.desktop file thats used for autostart but, IMHO, the %f should be ignored by the launcher if it, the launcher, has nothing to put in there.  

Its possible that the launcher is literally passing %f to tomboy-ng, as a non switch parameter, tomboy-ng assumes you want to open a file by that name. Not finding one, it opens a new one. Thats all exactly as it should be. Except the %f should not be passed through ....

Alternativly, MX is actually putting something on the command line.  When this happens, can you see, in the note's title bar, what the name of the new note is ?  And what the title of the note is (ie the blue text to of body of the note) ?

Davo

**Comment by aguador** [comment]
tomby-ng opens a new note with the default "New Note [Date and Time]" as title and name.

**Comment by aguador** [comment]
Yes, that will do it. However, I also noted in editing that there were several spaces after "%f" for whatever reason. Simply eliminating those spaces solves the problem as well.

There is another issue unaffected for which, unless you tell me it is expected behaviour, I will open a separate report: if you close session with a note open, then with autostart in the new session the note is reopened.

**Comment by aguador** [comment]
Have done further testing as trailiing spaces, if really present should be ignored. I am now not able to reproduce what was consistently (bad) behaviour before. tomboy-ng is autostarting properly from its config file with or without %f. The only thing I have done in the interim was to test with a light theme. However, even changing back to the dark theme does not reproduce the bug. Perhaps something was simply corrupted in the config file?

Will close this in a couple of days if I canno reproduce it again.

**Comment by davidbannon** [comment]
Eh ?  Now that is weird indeed.  Lets get this right, step by step. (Just checked the origional desktop file, there is only one space between the 'tomboy-ng' and '%f'. No trailing spaces.)

1. You edited the desktop file, removed the %f and, I guess any associated spaces ? You found additional spaces between the two words ?  And that worked OK.

2. You then put the %f back in and it still works ?

I don't rewrite that desktop file, just copy it. I wonder if MX somehow plays with it, maybe ?  Cannot imagine why.  Multiple spaces should be ignored, its "whitespace" but its just possible their launcher does not check for whitespace correctly.

Do you think I should make a note of this under "known bugs" ?

ps - yes, an open note when system is shutdown should be re-opened on startup.

Davo

**Comment by aguador** [comment]
Forget the extra spaces bit as that was my stupidity. None there really.

However, yes. It worked properly after removing %f. I then put it back, and it still worked properly. I want to try purging -ng and configuration and reinstalling. The only thing I can imagine is that there was corruption in the installation somewhere.

What is so weird is that I played with -ng 7 ways to Sunday* and the error appeared consistently, then as soon as I touched the .desktop launcher it started behaving properly.

*OK, it's late here. I do not recall hearing that expression in decades or ever actually using it!

**Comment by aguador** [comment]
The results of further testing are nuts:

Procedure:

1. Purge tomboy-ng
2. Delete configuration file
3. Reinstall tomboy-ng
4. Change -ng settings to autostart
5. Close session & open a new one

Result: -ng opens a new note (without dark theme), upon closing the new note splash screen flashes on screen, -ng closes

6. Edit tomboy-ng.desktop to eliminate "%f"
7. Close session & open a new one

Result: -ng opens a new note (without dark theme), upon closing the new note splash screen flashes on screen, -ng closes - huh?
Double check with "ps aux" - shows "tomboy-ng %f" running - double huh??

8. Try changing to light theme (crazy, I know, but worth a try)
9. Close/open session

Result: the same. About ready to give up, but having tried one crazy idea . . .

10. Open tomboy-ng settings and change search to "Any combination"
11. Close/open session

Result: tomboy-ng splash appears (without dark theme); "ps aux" confirms the ovbious: "tomboy-ng" is running

Hypotheses:

1. Behaviour is related to quick startup that does not grab theme properly (#110), a change in other settings somehow affecting loading time ever so slightly (BUT a. theme still not picked up, and b. as I recall, original cfg was set to "Any combination" on search - from prior version of tomboy-ng)

2. Behaviour is related to some weird interactions in the settings

I have to admit that I have not seen anything like this, but there is obviously something wrong, at least with -ng in XFCE. (It would, perhaps, be nice to know if new XFCE 4.14 may behave better on theme and this issue.)

I am willing to play more if you have suggestions, but you may have to see if you can reproduce this in order to fix it. If it is possible to do a test version that does, despite difficulties, have a small sleep delay built in, I can test with respect to this bug as well.

**Comment by aguador** [comment]
Neglected something important: the test with the change in search preferences was with "EXEC=tomboy-ng %f" in the .desktop file.

**Comment by davidbannon** [comment]
Just to be sure here. We need, IMHO to keep these two issues separate, that is, tomboy-ng opening in single note mode in autostart AND secondly, the issue of it not picking up the desktop colours when in autostart.  While both affect autostart, I am sure they are quite different things.  Now, because I have been distracted recently but able to do some thinking today while sinking some fence posts -
1. You have only seen both issues on MX Linux ?
2. Does removing the "%f" from the desktop file in autostart always solve the single note mode issue ?
3. On MX, does right clicking a tomboy note and choosing tomboy-ng open the note (in single note mode) ?

Maybe I make a small app that just reads its command line and displays it ?  You could manually put a desktop file in autostart and try that ?  I have been repeating your exercise and did find the autostart OK but U19.10 had some colours wrong if I set a dark theme before restarting. (quite an issue setting a dark theme too, need to install gnome-tweaks). 
Davo

**Comment by davidbannon** [comment]
[tester.zip](https://github.com/tomboy-notes/tomboy-ng/files/3754243/tester.zip)
OK, small test programme.  Unzip and edit the .desktop to reflect where you have put the executable. Then put the desktop file in ~/.config/autostart/. and see what you get.

Its also got several unusable controls on it to see what colour they come out as.

**Comment by aguador** [comment]
Setting fence posts is good for the soul. However, I have not checked the moon though, so be careful. The great uncle of a friend claimed that if you set posts when the crescent was pointing upward the posts would work their way up out of the ground.

> 1. You have only seen both issues on MX Linux ?

Yes, Mageia is OK.

> 2. Does removing the "%f" from the desktop file in autostart always solve the single note mode issue ?

No, note the results after steps 6 & 7 above. It is not until after making a change in the settings/cfg file that the issue rights itself - then it does not matter if the %f variable is there or not.

> 3. On MX, does right clicking a tomboy note and choosing tomboy-ng open the note (in single note mode) ?

Not 100% sure if this is what you wanted. However, without tomboy-ng running I went to the notes directory, right clicked and chose open with tomboy-ng. The note correctly opened and closed in single-note mode.
> 
Your test file autostarts leaving a "splash" window on the desktop that does NOT pick up the dark theme (1st png). It also adds the -ng icon to the panel. Clicking on that icon opens the "spalsh" using the dark theme (2nd png). Note the differences in the upper panes of the two windows.

![Auto-test-Form1](https://user-images.githubusercontent.com/8114758/67275494-0b7c3400-f4c3-11e9-8574-93c18114eead.png)
![Auto-test-Form1_panel](https://user-images.githubusercontent.com/8114758/67275502-0fa85180-f4c3-11e9-9622-ae75d7406923.png)

**Comment by davidbannon** [comment]
OK, so, two answers there. Firstly, looking at dark theme, the first image you posted show it partly converted to dark. That is pretty clearly related to what we thought, MX is starting autostart app before it has sorted out its own colours. Only solution I see if to make the autostart start something else, something that delays for a second or so and then it starts tomboy-ng. Thats a workaround more than a fix internal to tomboy-ng. Other platforms don't need or want it.

And the first image clearly shows the "%f" being passed through to the app as a parameter. Again, thats an MX issue IMHO, its launcher should not do that. I can deal with that reasonably easily however, before I jump to single note mode, I have a look at the parameter, if its "%f", ignore it. That won't have any ill effects on other platforms.
I'll cut you a binary .....

**Comment by davidbannon** [comment]
[tomboy-ng.zip](https://github.com/tomboy-notes/tomboy-ng/files/3758902/tomboy-ng.zip)

OK, here is a new binary that checks to see if the passed parameter is not %f before declaring it a single note mode thingo. Unzip and you will have to put it in place of the existing installed binary as root. Its /usr/bin/tomboy-ng I expect ....

The moon is fairly bright at the moment, I'll have a look tonight to see if crescent tips are up or down. if my posts do pop out, at least now I will have an excuse....

Davo

**Comment by aguador** [comment]
OK, I finally got time to test the file. Yes, tomboy-ng starts in "normal" mode, not single note. (See screenshot.) However, poking around a bit more:

- Using tomboy-ng's autostart feature, the running process is shown as: "tomboy-ng %f"

- Using XFCE's autostart (as I normally do), rather than -ng's, the running process is "tomboy-ng"

And it gets better. In Mageia 7 where I have release 0.23 so I can run it in background, it is autostarted by Enlightenment. If I try to start it with the -ng autostart feature . . . well . . . it's a non-starter, I presume because E controls autostart, not /etc/xdg/autostart. Would results be similar in a few other DEs or WMs? And if we ever get there, Wayland?

Aside: MX Linux 19 is out, with XFCE 4.14. I need to play with doing a clean install before dealing with a couple of novices I have running MX. I hope I will be able to play a bit this week and see if -ng is OK with the lastest XFCE.
![New-tomboy-ng](https://user-images.githubusercontent.com/8114758/67584866-d2f08a80-f74e-11e9-923e-93c4b219038d.png)

**Comment by davidbannon** [comment]
Firstly, I would rather you did not mention Wayland in polite company.

II think that "ignore a note called %f" fix will be there at next release.

Davo

**Comment by davidbannon** [comment]
Next release out soon will have code that rejects a file called %f so, while it should not be passed to the app, it won't matter if it is.
